### ISSUE TITLE
Intercepting events in a class implementing EventDispatcher

### ISSUE THREAD

**Issue by eterevsky** [pyglet/pyglet#205]
## TL;DR

I would like to add an ability for the class inheriting from `EventDispatcher` to add high-priority handlers, in addition to low-priority handlers that they have now.

## Current behavior

Currently the logic of dispatching events in the `EventDispatcher` is as follows:

* Go through `self._event_stack` and call the handlers until one returns `EVENT_HANDLED`.
* If none of the handlers returned `EVENT_HANDLED`, call the method with the same name on the class itself.

## The problem

I need to first call the handler on the class, and then call the external handlers.

My use case is as follows: I'm trying to implement a simple UI (see [current version](https://github.com/eterevsky/seer/blob/2551d3ab2613f080ab4473249f7d46861b0c517e/ui.py)). In particular I have a layout class `StackLayout` that owns some `Pane` objects and routes events to them. Then there's code that listens to events on `Pane` objects and does the actual rendering. The problem is that that application code is called *before* `Pane.on_event`, and it means for example that when the handler for `on_resize` is called, the `Pane` object fields are not updated yet with the new size.

It seems that the current logic of `EventDispatcher` was written under the assumption that all the events of the class are raised by the code in it, and `DispatcherClass.on_event` methods are just fallback handlers. But in my case it's more convenient to raise the events from the parent class.

## Proposal

In my mind the cleanest solution for this would be to add optional decorators for the event handlers that would classify them as either "default" (or maybe "fallback") or "intercept":

```python
    @events.intercept
    def on_draw(self, ...):
      # ...

    @events.default  # Optional. Same behavior as now.
    def on_resize(self, ...)
      # ...

    @events.intercept('on_resize')
    def on_resize_intercept(self, ...):
      # ...
```

The intercept handler will be called *before* the main event stack, and the default handler will be called *after*, provided that all handlers have returned `EVENT_UNHANDLED`.

This behavior could optionally be generalized to the external event listeners by introducing the second event stack (or more precisely queue), that would grow in the opposite direction, so that the handlers are called in the following order `intercept_handler1`, `intercept_handler2`, `intercept_handler3`, `default_handler3`, `default_handler2`, `default_handler1`, where the handler with the lower number was registered earlier:

```python
    @dispatcher.event(intercept=True)
    def on_resize(width, height):
        # ...
```

or

```python
    @dispatcher.event
    @events.intercept
    def on_resize(width, height):
        # ...  
```

## Alternatives

Here are some alternatives that I considered, that would not require the changes to `EventDispatcher`. I wasn't happy with any of them, but maybe I missed some other solution.

* Dispatch the events on the parent class, and listen to them in the child classes. (In my code, the `Pane` object will listen to the events on the parent `StackLayout` object.) This restores the invariant that the events on an object are only raised by itself. But in my case it means that every child would listen to every event on the parent. This prevents any possible optimizations of event-routing in the parent object, and requires an extra layer of events (in my current code only `Pane` is an `EventDispatcher`, but not `StackLayout`).

* Add an extra method on the child class for every possible event and just call it explicitly from the parent instead of raising the event. This will require more boilerplate, especially for the events that `Pane` otherwise doesn't need to handle at all like `on_mouse_press`.

* Just require the application code of the widget that happens to be on the top of the handlers stack to call the code in the object that needs to be called.

## Additional observation

* Currently `dispatch_event` acts as if the class just called `self.push_handlers(self)` before any other handlers. But for some reason, the handlers on the instance itself are called separately. Would it make sense just to add the instance to `_event_stack`, and remove special handling?

* Some comments in `events.py` mention the method `EventDispatcher.dispatch_events`, which as far as I can tell is defined in `BaseWindow`. I am not sure if this method is expected to be implemented for other classes implementing `EventDispatcher`, but perhaps the comments should be updated.

**Comment by benmoran56** [comment]
Thanks for opening the ticket, and for the suggestion.

For the EventDispatcher subclasses, you pointed out how the subclass methods are called separately and not just added to the `_event_stack`. I think adding them to the stack would be a good idea, and would be a first step to potentially opening up sorting the stack. I had a look at this, but I couldn't find a way to make it work with multiple subclasses. If you have any suggestions, that would be great!

Your proposal looks reasonable. One thing to note is that the decorator currently replaces the top handler on the stack, whereas the `push_handler(s)` methods will actually stack them. Maybe this behavior can also be reevaluated.

**Comment by eterevsky** [comment]
@benmoran56, thanks for the comments! I didn't notice that the decorator works differently from `push_handler()`. I think it is worth rethinking, but it should be done with care, since this change could break existing code. This also means that until the semantics of the decorator is changed, we can't just push the class onto `_event_stack`, since it will be replaced by the decorators, which doesn't happen now.

One other thing that just came to mind regarding `self.push_handlers(self)`: what should happen if the user just overwrites a handler method after this, like

```python
dispatcher.on_event = lambda _: EVENT_HANDLED
```

The tutorial allows this, but this wouldn't work if we call the handlers on the dispatcher via the event stack.

Let me quickly make a PR to add intercepting handlers without changing any of the present behavior, and then you can decide whether you want to make the breaking changes.

**Comment by benmoran56** [comment]
For breaking changes, the policy has been to deprecate changes for one release cycle, then remove in the next.  Currently, however, we're in a bit of a transitional period since pyglet 1.5.x is likely to be the last release of pyglet that supports OpenGL 2.0. The upcoming pyglet 2.0 is being ported to OpenGL 3.3+, and will be breaking the API in many places out of necessity. We are planning to maintain 1.5 in tandem with 2.0 for the near future, as long as their is a demand. Since it's the last legacy release, we want to make sure it works as expected for existing code bases. 

So while 2.0 is not yet out (only dev releases so far), it is an opportunity to make breaking changes where it make sense.

**Comment by eterevsky** [comment]
I've just played around with the event module, and it seems that it can be made much simpler by only supporting "push" semantics and removing the event "frames" altogether. This will allow maintaining a queue of handlers separately for each event type. Instead of iterating over handlers like this:

```python
for frame in self._event_stack:
  handler = frame.get(event_type, None)
  if handler is not None:
    # ...
```

we'd be able to iterate like this:

```python
for handler in self._event_stack.get(event_type, ()):
  # ...
```

which should be more efficient, as well as simpler.

This change would only be possible if `set_handlers()` and `pop_handlers()` are removed, since they rely on the notion of the "current" events frame. I'm not sure how widely used are these functions, and why would anyone replace the current handler at all.

**Comment by eterevsky** [comment]
Here's a pull request that adds the intercepting handlers without breaking the 1.5.x semantics: #207.

If you wish, I could also implement the breaking change for 2.x that we've discussed above.

**Comment by caffeinepills** [comment]
While I like this idea and have been wanting something like it for a while. I have a feeling in the back of my mind that this is a bandaid fix for what is essentially a problem I've viewed with the event system; there is not much control or visibility about the stack. Generally you just push events and if the order becomes messed up, you have to just wipe the stack and recreate it manually to your liking.

A question: What if I intercept two of the same events, which runs first? Does one overwrite the other? If the former, does that become a stack in the stack?

There is also a use case I had when I wanted an event to be called very last, but the problem became if I added other events, then there is no way to keep it at the end. IMO for a true priority system we should have access to insert at any point in the stack, view the stack, prioritize events (like this PR) (for example, if I insert at 0, I should be able to specify I want this to stay at 0 to always be called first, or -1, to always be called at the end).

**Comment by eterevsky** [comment]
@caffeinepills I thought about explicit priorities for event handlers, but in my experience, it will unnecessarily complicate things. The problem is, that it doesn't solve the problem that you mention. Suppose two handlers both specify priority -1, wanting to be called at the end. Which one will be called later?

In my proposal there is a deterministic order between all handlers: the handler with `@intercept` are invoked in the order of registration, and normal handler are invoked in the opposite order (same as now). So the dispatcher itself can register the very first and the very last handlers. Intuitively this convention reflects a sequence of nested calls: the higher-level function execute code nearer to the beginning and end of the execution.

**Comment by caffeinepills** [comment]
> The problem is, that it doesn't solve the problem that you mention. Suppose two handlers both specify priority -1, wanting to be called at the end. Which one will be called later?

The same way yours would handle it.,

> invoked in the order of registration

**Comment by eterevsky** [comment]
> The same way yours would handle it.,

In that case why do you need explicit priorities?

**Comment by benmoran56** [comment]
For priorities, if the stack behaved like a heap (could probably *be* a heap internally), then you could do stuff like pushing a handler to any position in the stack at runtime. This could be nice if you want to do this at runtime, since you can't practically re-create the stack by adding in order.

**Comment by eterevsky** [comment]
@benmoran56 I didn't really want to add explicit priorities not because it's difficult to implement, but because it introduces unnecessary ambiguity when using events. Like, how would you determine what specific values you want to use? How do you ensure that an event handler is always invoked first?

In my proposal the programmer can choose whether their event handler is run before or after all the subsequently added handlers. It seems to be enough for any real-world situations.

**Comment by benmoran56** [comment]
The way I could see it working is like this: 

If all handlers default to `priority=0`, then they would be invoked in the order in which they were added. The priority is optional. If you wanted to add a specific handler that will be run before all others, you can give it a high priority (such as `priority=99`) and it will always be executed first, as long as you never add another handler with a higher priority of course. This would also make it possible to set a "first" and "last" handler, and later add/remove handlers in between them, if the need arose.

**Comment by eterevsky** [comment]
Just to note, the current order of handlers is the opposite: the last added handler is invoked first. I don't have any preferences about this, just wanted to point out that this will change the default behavior.

What will happen if two unrelated subsystems both want to catch the event first (or last)? Do they just compete by naming as high priority as they can?

Also, if my experience, when you allow such arbitrary priorities, it usually leads to a mess: different developers have different idea of what the priorities should be. Like, why do you use 99 in particular? At least, if you want numeric priorities, it's better to create some specific tiers and recommend users to use them rather than naked numbers:

```python
PRIORITY_FALLBACK = -1
PRIORITY_DEFAULT = 0
PRIORITY_OVERRIDE = 1
```

**Comment by benmoran56** [comment]
> Just to note, the current order of handlers is the opposite: the last added handler is invoked first. I don't have any preferences about this, just wanted to point out that this will change the default behavior.

Oops, yes you're right. In any case, it will be easy enough to maintain ordering when adding. 

> What will happen if two unrelated subsystems both want to catch the event first (or last)? Do they just compete by naming as high priority as they can?

If two handlers have the same priority, it will just fall back to the order in which they were added. 

> Also, if my experience, when you allow such arbitrary priorities, it usually leads to a mess: different developers have different idea of what the priorities should be. Like, why do you use 99 in particular? At least, if you want numeric priorities, it's better to create some specific tiers and recommend users to use them rather than naked numbers:
> 
> ```python
> PRIORITY_FALLBACK = -1
> PRIORITY_DEFAULT = 0
> PRIORITY_OVERRIDE = 1
> ```

I personally prefer integers, but this is a good usability point.

**Comment by eterevsky** [comment]
@benmoran56 If you agree with this API, I can change my PR to implement it. I have one question though: I can implement it in 1.5 branch or in 2.x branch. In 2.x I could also [simplify the handler queues](https://github.com/pyglet/pyglet/issues/205#issuecomment-633499163): get rid of the frames and just maintain a list of handlers per event.

I also need some implementation of priorities for my own pet projects, so if the 2.x version is currently unusable, I'd prefer to make the change in 1.5.

**Comment by benmoran56** [comment]
I mentioned the policy in a previous reply: https://github.com/pyglet/pyglet/issues/205#issuecomment-633492614

It would be best to target this for 2.0, since we can't make breaking changes in the 1.X series at this point. 

2.0 is usable for simple cases with Sprites and Text, but the rest of the API is not yet stabilized.

**Comment by eterevsky** [comment]
Heads-up: I rewrote the events framework according to the plan from [this comment](https://github.com/pyglet/pyglet/issues/205#issuecomment-633499163): I removed `set_handler`(s), event frames, changed the way events are stored. I've also updated the docs and unit tests to match the new implementation: https://github.com/eterevsky/pyglet/blob/events2/pyglet/event.py

One thing I didn't do yet is introduce the priorities. I hope to do it tomorrow, and will send a new PR then.
